### PR TITLE
Navigation common UI tests

### PR DESCRIPTION
**Pull request by terrakok**
Move a bunch of Navigation UI tests to common.

Fixes [CMP-6689](https://youtrack.jetbrains.com/issue/CMP-6689) iOS Navigation Crash Nested Graphs popBackStack

### Fixes - Navigation
- Fixed `No destination with ID 0 is on the NavController's back stack` crash on iOS.

### CODE REVIEW

File: navigation/navigation-common/src/commonMain/kotlin/androidx/navigation/NavDestination.kt [hunk 1]
```diff
@@ -102,6 +102,9 @@ public expect open class NavDestination(
     public open val displayName: String
         @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP) get
 
+    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+    public var id: Int
```
Comment (MatkovIvan) [inline navigation/navigation-common/src/commonMain/kotlin/androidx/navigation/NavDestination.kt]: I'd avoid exposing this to common
1. It doesn't make much sense for Compose
2. It messes up annotations on Android

I'd suggest to replace with expect/actual internal extension in tests

File: navigation/navigation-common/src/jbMain/kotlin/androidx/navigation/NavDestination.jb.kt [hunk 2]
```diff
@@ -124,7 +124,7 @@ public actual open class NavDestination actual constructor(
     }
 
     @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
-    public actual open fun matchDeepLink(route: String): DeepLinkMatch? {
+    public actual fun matchDeepLink(route: String): DeepLinkMatch? {
```
Comment (MatkovIvan) [inline navigation/navigation-common/src/jbMain/kotlin/androidx/navigation/NavDestination.jb.kt:127]: I know this place is a mess, but you're removing an ability to override it from public. Is it expected? It's restricted to `LIBRARY_GROUP`, so probably it's ok 🤔

File: navigation/navigation-compose/src/androidInstrumentedTest/kotlin/androidx/navigation/compose/NavBackStackEntryProviderTest.android.kt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021 The Android Open Source Project
+ * Copyright 2024 The Android Open Source Project
```
Comment (MatkovIvan) [inline navigation/navigation-compose/src/androidInstrumentedTest/kotlin/androidx/navigation/compose/NavBackStackEntryProviderTest.android.kt]: Please revert year in copyright for easier upstreaming

File: navigation/navigation-runtime/src/jbMain/kotlin/androidx/navigation/NavBackStackEntryState.jb.kt [hunk 4]
```diff
@@ -66,7 +66,7 @@ internal data class NavBackStackEntryState(
             if (bundle == null) return null
             val id = bundle.getString(KEY_ID) ?: return null
             val destinationId = bundle.getInt(KEY_DESTINATION_ID)
-            val args = bundle.getBundle(KEY_ARGS) ?: return null
+            val args = bundle.getBundle(KEY_ARGS)
```
Comment (MatkovIvan) [inline navigation/navigation-runtime/src/jbMain/kotlin/androidx/navigation/NavBackStackEntryState.jb.kt:69]: Was it a cause of CMP-6689?

**Review by MatkovIvan — COMMENTED** [review]
Thanks! It's a huge improvement in nav library port!

File: navigation/navigation-common/src/jbMain/kotlin/androidx/navigation/NavDestination.jb.kt (hunk 2 above)
Comment (terrakok) [reply navigation/navigation-common/src/jbMain/kotlin/androidx/navigation/NavDestination.jb.kt:127]: yes. I synchronized the API with the android part. there was a mistake. The change fixes the problem with a nested navigation
https://github.com/JetBrains/compose-multiplatform-core/blob/778cf002523049dfcf792198d3c223bbe3c181e2/navigation/navigation-common/src/androidMain/kotlin/androidx/navigation/NavDestination.android.kt#L355

Comment (terrakok) [reply navigation/navigation-common/src/jbMain/kotlin/androidx/navigation/NavDestination.jb.kt:127]: it should fix https://youtrack.jetbrains.com/issue/CMP-6689
I will check

File: navigation/navigation-runtime/src/jbMain/kotlin/androidx/navigation/NavBackStackEntryState.jb.kt (hunk 4 above)
Comment (terrakok) [reply navigation/navigation-runtime/src/jbMain/kotlin/androidx/navigation/NavBackStackEntryState.jb.kt:69]: no. this problem was found in one of the tests

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**